src/scala/microsoft-spark-2.4.x/src/main/scala/org/apache/spark/api/dotnet/CallbackConnection.scala

```diff
@@ -0,0 +1,143 @@
+ * @param address The address of the Dotnet CallbackServer
+ * @param port The port of the Dotnet CallbackServer
+ */
+class CallbackConnection(address: String, port: Int) extends Logging {
+  private[this] val socket = new Socket(address, port)
+  private[this] val inputStream = new DataInputStream(socket.getInputStream)
+  private[this] val outputStream = new DataOutputStream(socket.getOutputStream)
+
+  def send[T](
+    writeBody: DataOutputStream => Unit,
+    readBody: Option[DataInputStream => T]): CallbackResponse[T] = {
+    logInfo("Calling callback...")
+
+    try {
+      SerDe.writeInt(outputStream, CallbackFlags.CALLBACK)
+      writeBody(outputStream)
+      outputStream.flush()
+    } catch {
+      case e: Exception => {
+        logError("Error writing to stream.", e)
+        return CallbackResponse(ConnectionStatus.ERROR_WRITE, None)
+      }
+    }
+
+    val readBodyResponse: Option[T] =
+      try {
+        readBody match {
+          case Some(body) => {
+            val returnValueFlag = checkForDotnetException(inputStream)
+            if (returnValueFlag != CallbackFlags.CALLBACK_RETURN_VALUE) {
+              throw new Exception("readBody defined, however flag to indicate return value not " +
+                s"received. Expected: ${CallbackFlags.CALLBACK_RETURN_VALUE}, " +
+                s"Received: $returnValueFlag")
+            }
+            Some(body(inputStream))
+          }
+          case None => None
+        }
+      } catch {
+        case e: Exception => {
+          logError("Error reading stream while checking callback return value.", e)
+          return CallbackResponse(ConnectionStatus.ERROR_READ, None)
+        }
+      }
+
+    logInfo(s"Signaling END_OF_STREAM.")
+    try {
+      SerDe.writeInt(outputStream, CallbackFlags.END_OF_STREAM)
+      outputStream.flush()
+
+      val endOfStreamResponse = checkForDotnetException(inputStream)
+      endOfStreamResponse match {
+        case CallbackFlags.END_OF_STREAM =>
+          logInfo(s"Received END_OF_STREAM signal. Calling callback successful.")
+          return CallbackResponse(ConnectionStatus.OK, readBodyResponse)
+        case _ =>  {
+          logError(s"Error verifying end of stream. Expected: ${CallbackFlags.END_OF_STREAM}, " +
+              s"Received: $endOfStreamResponse")
+        }
+      }
+    } catch {
+      case e: Exception => {
+        logError("Error while verifying end of stream.", e)
+      }
+    }
+
+    CallbackResponse(ConnectionStatus.ERROR_END_OF_STREAM, readBodyResponse)
+  }
+
+  def close(): Unit = {
+    try {
+      SerDe.writeInt(outputStream, CallbackFlags.CLOSE)
+      outputStream.flush()
+    } catch {
+      case e: Exception => logInfo("Unable to send close to .NET callback server.", e)
+    }
+
+    close(socket)
+    close(outputStream)
+    close(inputStream)
+  }
+
+  private def close(s: Socket): Unit = {
+    try {
+      if (s != null) {
```

Comment:
isn't this assert? i.e., in what places do you intend to change socket to null in this class?